### DISCUSSION TITLE
The rewind simulation is inconsistent

### DISCUSSION THREAD

**Discussion by c068162** [bepu/bepuphysics1#49] [Q&A] [answered]
Sorry for my poor English, please ask me what state of synchronization should be required for rewind. At present, I use BepuInt version in Unity and server simultaneously, and the synchronization method I use is Rocket Alliance's full prediction scheme, everything is normal when there is no rewind, but once the rewind happens after physical simulation, There will be errors on the client side and the server side, and my rewind is just a re-simulation of the input. May I ask if there is a solution, or what part of me has a problem? I would be grateful if you could answer me。
This is my rewind code：
`entity.bepuEntity.Position = new Vector3(lsfMoveCmd.PosX, lsfMoveCmd.PosY, lsfMoveCmd.PosZ);
entity.bepuEntity.Orientation = new Quaternion(lsfMoveCmd.RotX, lsfMoveCmd.RotY, lsfMoveCmd.RotZ, lsfMoveCmd.RotW);
entity.bepuEntity.LinearVelocity = new Vector3(lsfMoveCmd.LinearVelocityX, lsfMoveCmd.LinearVelocityY, lsfMoveCmd.LinearVelocityZ);
entity.bepuEntity.AngularVelocity = new Vector3(lsfMoveCmd.AngularVelocityX, lsfMoveCmd.AngularVelocityY, lsfMoveCmd.AngularVelocityZ);`

This is my simulation code：
`self.bepuEntity.ApplyImpulse(self.bepuEntity.Position, self.Input * 18m * deltaTime);`

**Comment by RossNordby** [comment]
The simulation has internal state which affects the result, like broad phase tree topology, constraint order, and solver accumulated impulses. Positioning objects at an earlier pose/velocity does not reset the other internal state.

A couple of options:
1. Implement an internal state deep copy operation. This isn't built-in and would be annoying.
2. Every frame, rebuild the simulation from scratch in the exact same order (like making a new simulation and adding everything at the current poses into it) so that the internal states are the same when you do a rebuild for a rewind. This carries significant overhead for large simulations.

There's no super easy and cheap solution, unfortunately.

**Reply by c068162** [reply]
Sorry, I found that I thought too simply. As you said, repositioning the position and speed does not reset other internal states. What I understand about option 2 is to resimulate from the first frame. This works for my game. It seems unlikely. It would be fine if we just go back to the frame where the error occurred and re-simulate. However, I tested and found that even the re-simulation in the same sequence would not return to the correct state, which would cause the error to occur from this frame onwards. is wrong, can you tell me what should be copied for the deep copy of option 1, do they have these properties in the Entity? thank you

**Reply by c068162** [reply]
Through Debug, I found that the value of the InertiaTensor attribute is inconsistent after resetting the state. I want to synchronize the InertiaTensor attribute, but the InertiaTensor cannot be assigned a value. Is there any way to assign a value, or change it indirectly?

**Reply by RossNordby** [reply, marked as answer]
The inertia tensor is just the rotated local inertia tensor. If it doesn't match, it's because either something changed the local inertia tensor, or (more likely) the orientation doesn't match.

> What I understand about option 2 is to resimulate from the first frame. 

I don't mean to resimulate from the first frame, but rather to rebuild the simulation using the most recent state. Rebuilding the simulation means that all internal state differences accumulated over updates are wiped out.

To phrase it another way, imagine 3 cases:
1. Simulation A starting from initial conditions and running for 10 frames.
2. Simulation B that is built from all the externally visible state (like entity poses/velocities) in A after the 10 frames.
3. Simulation C that is also built from all the externally visible state in A after the 10 frames.

B and C would have internal state that is *not* the same as A. Continuing to run the simulation would result in B and C diverging from A, despite having the same externally visible states.

However, B and C would share the same internal state, and deterministically continuing the simulation would result in them continuing to share the same state.

In other words, rebuilding the simulation wipes out differences internal state and reconstructs it in a predictable way. If, for every single frame, you reconstruct the simulation in this way, and *also* do that when doing your rewind simulation, then the external states fully define what the internal states are at each step and they will remain consistent.

> can you tell me what should be copied for the deep copy of option 1, do they have these properties in the Entity? thank you

The order of all constraints, the order of nodes in the broad phase, accumulated impulses in the solver, cached collision information, and so on. It's not just entity states; it's a ton of different pieces that would be really, really annoying to copy.

**Reply by c068162** [reply]
I think I understand what you mean. It sounds great. Every frame, create a new Space and Entity at TickStart, assign the snapshot of the previous frame to the newly created Entity, and then refresh the physical world in Tick. Finally Remove Space and Entity at TickEnd. I don't know if the performance overhead of doing this is very large, even though I only have 2-6 Boxes in each physics world. Thank you very much for your suggestion, I will adopt this solution, and finally wish you a happy life.

**Reply by c068162** [reply]
This solution was successful. Now everything is deterministic. The performance of this method is very poor. I found that the GC is a bit high. How should I destroy Space and Entity correctly?

**Reply by RossNordby** [reply]
In v1, nothing special; toss them out the window and the GC deals with it. It does indeed cause a lot of garbage, and it will be slower due to the rebuild. The GC costs can be mitigated some by keeping the rest of the [heap complexity](https://shawnhargreaves.com/blog/twin-paths-to-garbage-collector-nirvana.html) low, but that isn't always practical.

**Reply by c068162** [reply]
Well, it seems that I can only optimize it slowly. You are really helpful to me. Thank you very much.

**Comment by c068162** [comment]
Thank you very much for your quick reply, and you are right. According to your second suggestion, I found that the execution sequence of each Tick foreach is different (applying pulse and recording snapshot). After I fixed the execution sequence of foreach, this problem was solved, and my FixedUpdate was 24Tick. Space is 120Tick, in order to optimize the network delay, I added the function of variable speed, but dt in space.update (dt) will not change. At first, I thought it was the problem here, but now everything seems to be smooth, and the local prediction will be 100% successful when the other party does not operate, which makes me very happy. Physical network synchronization is an interesting thing, thank you for your help to me, wish your physics engine can develop better and better, and finally wish you a happy life. 😀
